plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/ui/contentAssist/ContentAssistFragment2.xtend

```diff
@@ -0,0 +1,297 @@
+/*******************************************************************************
+ * Copyright (c) 2015 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.xtext.generator.ui.contentAssist
+
+import com.google.common.collect.Sets
+import com.google.inject.Inject
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.xtend.lib.annotations.Accessors
+import org.eclipse.xtend2.lib.StringConcatenationClient
+import org.eclipse.xtext.AbstractElement
+import org.eclipse.xtext.AbstractRule
+import org.eclipse.xtext.Alternatives
+import org.eclipse.xtext.Assignment
+import org.eclipse.xtext.CrossReference
+import org.eclipse.xtext.Grammar
+import org.eclipse.xtext.RuleCall
+import org.eclipse.xtext.xtext.generator.AbstractGeneratorFragment2
+import org.eclipse.xtext.xtext.generator.CodeConfig
+import org.eclipse.xtext.xtext.generator.XtextGeneratorNaming
+import org.eclipse.xtext.xtext.generator.model.FileAccessFactory
+import org.eclipse.xtext.xtext.generator.model.GuiceModuleAccess
+import org.eclipse.xtext.xtext.generator.model.TypeReference
+
+import static extension org.eclipse.xtext.GrammarUtil.*
+import static extension org.eclipse.xtext.xtext.generator.util.GrammarUtil2.*
+
+/**
+ * Contributes the 'Abstract...ProposalProvider' and '...ProposalProvider' stub,
+ *  the latter either in Xtend or Java language.
+ * 
+ * @author Christian Schneider - Initial contribution and API
+ */
+class ContentAssistFragment2 extends AbstractGeneratorFragment2 {
+
+	@Inject
+	extension XtextGeneratorNaming
+	
+	@Inject
+	extension CodeConfig
+	
+	@Inject
+	FileAccessFactory fileAccessFactory
+
+	@Accessors
+	boolean generateStub = true;
+
+	@Accessors
+	boolean inheritImplementation = true
+
+	def protected TypeReference getProposalProviderClass(Grammar g) {
+		return new TypeReference(
+			g.eclipsePluginBasePackage + ".contentassist." + g.simpleName + "ProposalProvider"
+		)
+	}
+
+	def protected TypeReference getGenProposalProviderClass(Grammar g) {
+		return new TypeReference(
+			g.eclipsePluginBasePackage + ".contentassist.Abstract" + g.simpleName + "ProposalProvider"
+		)
+	}
+
+	def protected TypeReference getGenProposalProviderSuperClass(Grammar g) {
+		val superGrammar = g.nonTerminalsSuperGrammar
```

Comment:
Agree. Actually I'm a fan of explicit return statements, but I wasn't sure whether you guys would use it in these cases.
Thanks for that remark :-)